[FEATURE] Add support for multiple cache implementations, separating result, query and metadata caching.

BrandonLegault opened the pull request:
This PR adds the ability to configure a different cache driver for each of metadata, result and query caching through the `doctrine.php` config file `'cache'` array.

### Changes proposed in this pull request:

The current cache array format is as follows:

```
'cache'            => [
    'default'      => env('DOCTRINE_CACHE', 'array'),
    'namespace'    => null,
    'second_level' => false,
]
```

This branch introduces this format:

  ```
    'cache' => [
        'second_level'     => false,
        'default'          => env('DOCTRINE_CACHE', 'array'),
        'namespace'        => null,
        'metadata'         => [
            'type'         => 'file',
            'namespace'    => null,
        ],
        'query'            => [
            'type'         => 'redis',
            'namespace'    => null,
        ],
        'result'           => [
            'type'         => 'array',
            'namespace'    => null,
        ],
    ],
```
The configuration is currently applied in this PR by calling `\Doctrine\ORM\Configuration->set[Result|Query|Metadata]CacheImpl` after reading the values from the config.

The `default` cache array entries remain the same to ensure backwards compatibility. If any of the metadata, result or query entries are missing from the configuration then they will take whichever the default entry is. The default entry is also mirrored/stored directly in code in case it's accidentally deleted from the configuration.

BrandonLegault committed:
Implement multiple cache driver support in EntityManagerFactory

BrandonLegault committed:
Make all EntityManagerFactory tests pass (may not be correct yet)

BrandonLegault committed:
Refactor the EntityManagerFactory to properly make use of cache drivers and second level caching

BrandonLegault committed:
Fix unit tests for EntityManagerFactory to reflect new EMF code

BrandonLegault committed:
Fix the default cache implementation to ensure backwards compatibility

BrandonLegault committed:
Fix formatting for StyleCI build

BrandonLegault committed:
Add more formatting for StyleCI

BrandonLegault committed:
Add even more StyleCI fixes

patrickbrouwers commented on src/EntityManagerFactory.php:
```diff
@@ -318,13 +325,27 @@ protected function setCustomFunctions(Configuration $configuration)
         $this->setSecondLevelCaching($configuration);
     }
 
+    private function applyNamedCacheConfiguration($cacheName)
```
Docblocks

patrickbrouwers commented on src/EntityManagerFactory.php:
```diff
@@ -318,13 +325,27 @@ protected function setCustomFunctions(Configuration $configuration)
+        $defaultDriver    = $this->config->get('doctrine.cache.default', $this->defaultCache['type']);
+        $defaultNamespace = $this->config->get('doctrine.cache.namespace', $this->defaultCache['namespace']);
+
+        $driverType = $this->config->get('doctrine.cache.' . $cacheName . '.type', $defaultDriver);
```
`type` should be named `driver`

patrickbrouwers reviewed (changes requested):
Can you also update the config file with the new structure, thanks!
It would be good to provide env variables for the 3 drivers as well. They can all 3 default to the DOCTRINE_CACHE env variable.

BrandonLegault committed:
Change multicaching to use driver instead of type in the cache array

Also added some docblock comments

BrandonLegault committed:
StyleCI will be the death of me

BrandonLegault commented:
I added the requested changes... I also rebased from 1.2 to keep my branch up to date. I'd already made the changes before I saw the DOCTRINE_CACHE env variable comment. I made them take on new env variables. Do you want me to change that?

patrickbrouwers commented on config/doctrine.php:
```diff
@@ -154,10 +154,22 @@
-        'second_level' => false,
+    'cache' => [
+        'second_level'     => false,
+        'default'          => 'array',
```
'default'          => env('DOCTRINE_CACHE', 'array'),

patrickbrouwers commented on src/EntityManagerFactory.php:
```diff
@@ -318,14 +325,32 @@ protected function setCustomFunctions(Configuration $configuration)
+        $defaultDriver    = $this->config->get('doctrine.cache.default', $this->defaultCache['type']);
+        $defaultNamespace = $this->config->get('doctrine.cache.namespace', $this->defaultCache['namespace']);
+
+        $driver = $this->config->get('doctrine.cache.' . $cacheName . '.driver', $defaultDriver);
```
```php
$cache = $this->cache->driver($driver);

if ($namespace = $this->config->get('doctrine.cache.' . $cacheName . '.namespace', $defaultNamespace)) {
    $cache->setNamespace($namespace);
}

return $cache;
```

patrickbrouwers commented on src/EntityManagerFactory.php:
```diff
@@ -318,14 +325,32 @@ protected function setCustomFunctions(Configuration $configuration)
+     */
+    private function applyNamedCacheConfiguration($cacheName)
+    {
+        $defaultDriver    = $this->config->get('doctrine.cache.default', $this->defaultCache['type']);
```
Replace `$this->defaultCache['type']` by `'array'`
Remove `$this->defaultCache['namespace']`

patrickbrouwers commented on src/EntityManagerFactory.php:
```diff
@@ -58,6 +58,14 @@ class EntityManagerFactory
     /**
+     * @var array
+     */
+    private $defaultCache = [
```
Not necessary

patrickbrouwers commented:
>  I made them take on new env variables. Do you want me to change that?

No it's fine.
